include/pybind11/pybind11.h

```diff
@@ -385,9 +384,22 @@ class cpp_function : public function {
                         std::string(pybind11::str(rec->scope.attr("__name__"))) + "." + std::string(rec->name) + signature
                     #endif
                 );
-            while (chain->next)
-                chain = chain->next;
-            chain->next = rec;
+
+            if (rec->is_prepended) {
+                // Beginning of chain; we need to replace the capsule's current head-of-the-chain
+                // pointer with this one, then make this one point to the previous head of the
+                // chain.
+                chain_start = rec;
+                rec->next = chain;
+                auto rec_capsule = reinterpret_borrow<capsule>(((PyCFunctionObject *) m_ptr)->m_self);
+                rec_capsule.set_pointer(rec);
```

Comment:
I feel like this is a little simpler, and providing access to `PyCapsule_SetPointer` is a small plus?

Reply:
(To be honest, I believe @jagerman added these details to the original PR, and I tried to do something more like your suggestion at first)

Reply:
Fair enough as well. Maybe it's better to not change the contents of an existing struct, yes. Seems `PyCapsule_SetPointer` is reasonably cheap anyway (and not really more expensive than copying the contents):

https://github.com/python/cpython/blob/1ed54435268b285964141fb74d47ceaa33ea79ab/Objects/capsule.c#L133-L149

(though, since it can throw errors, etc, you do need to hold the GIL to call it, ofc, but we do hold the GIL already at that point)

Reply:
Apart from that:
1. It might be a bit weird to only expose `set_pointer`, when there are so many other `PyCapsule_` functions? But they can be added when necessary. Alternatively, we don't expose that as public API and just use `PyCapsule_SetPointer` inernally?
2. Given that `PyCapsule_GetPointer` is disguised as `operator T*()`, should this then be `operator=(T*)`?

Reply:
1. So many? I think it's basically four pairs, is it not? Get/set pointer, name, destructor, and context? And we already provide some of that.
2. I think that would be fine, as long as someone doesn't expect the destructor to be called when setting it?